src/finch/finchassembly/nodes.py

```diff
@@ -0,0 +1,373 @@
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Any
+from ..symbolic import Term
+
+
+@dataclass(eq=True, frozen=True)
+class AssemblyNode(Term):
+    """
+    AssemblyNode
+
+    Represents a FinchAssembly IR node. FinchAssembly is the final intermediate representation
+    before code generation (translation to the output language). It is a low-level imperative
+    description of the program, with control flow, linear memory regions called "buffers", and
+    explicit memory management.
+    """
+
+    @staticmethod
```

Comment:
Hmm. Maybe it makes sense to do `class Expression(Term)` and change `LogicExpression` to inherit from both `LogicNode` and `Expression`.

Reply:
I'd like to leave this alone for now, and have a discussion about it in a separate issue. There are many very fancy ways we can automate IR definitions, but I like that the current design is really simple for now.

Reply:
Separate issue here: https://github.com/finch-tensor/finch-tensor-lite/issues/23